### PR TITLE
Enable strict compiler settings

### PR DESCRIPTION
**Pull request by bpkroth**
This change enforces javac compiler warnings and linting as errors as a part of #406.

Splitting work out from #410, though that should be merged first, and may require additional fixups after that (e.g., new compiler constraints).

### CODE REVIEW

File: src/main/java/com/oltpbenchmark/benchmarks/wikipedia/data/RevisionHistograms.java [hunk 1]
```diff
@@ -290,12 +290,16 @@ public abstract class RevisionHistograms {
      */
     public static final int[] REVISION_DELTA_SIZES = {1000, 10000, 100000};
 
+    /**
+     * Simple generic class overload to avoid some cast warnings below.
```
Comment (anjagruenheid) [inline src/main/java/com/oltpbenchmark/benchmarks/wikipedia/data/RevisionHistograms.java:294]: Should this type of handling be part of a best practices doc? I.e., avoid casting, use some structures like this instead.

File: src/test/java/com/oltpbenchmark/benchmarks/auctionmark/TestAuctionMarkWorker.java [hunk 2]
```diff
@@ -61,7 +60,7 @@ public void testUniqueSellers() throws Exception {
         Set<UserId> all_users = new HashSet<UserId>();
         Set<UserId> worker_users = new TreeSet<UserId>();
         Integer last_num_users = null;
-        for (Worker w : this.workers) {
+        for (var w : this.workers) {
```
Comment (anjagruenheid) [inline src/test/java/com/oltpbenchmark/benchmarks/auctionmark/TestAuctionMarkWorker.java:63]: Why the change?

File: src/main/java/com/oltpbenchmark/benchmarks/auctionmark/AuctionMarkLoader.java [hunk 3]
```diff
@@ -1195,6 +1195,7 @@ protected class ItemBidGenerator extends SubTableGenerator<LoaderItemInfo> {
         private LoaderItemInfo.Bid bid = null;
         private float currentBidPriceAdvanceStep;
         private long currentCreateDateAdvanceStep;
+        @SuppressWarnings("unused") // spurious
```
[the remark this answers is not in the document]
Comment (bpkroth) [reply src/main/java/com/oltpbenchmark/benchmarks/auctionmark/AuctionMarkLoader.java:1198]: Without it `currentPrice` get's an alert about being unused (it's only ever assigned, not read/returned).

File: src/main/java/com/oltpbenchmark/benchmarks/wikipedia/data/RevisionHistograms.java (hunk 1 above)
Comment (bpkroth) [reply src/main/java/com/oltpbenchmark/benchmarks/wikipedia/data/RevisionHistograms.java:294]: Not a bad idea.  It's basically a limitation in the java language spec as I understand it.  You can't create arrays of a generics:
https://docs.oracle.com/javase/tutorial/java/generics/restrictions.html#createArrays

File: src/test/java/com/oltpbenchmark/benchmarks/auctionmark/TestAuctionMarkWorker.java (hunk 2 above)
Comment (bpkroth) [reply src/test/java/com/oltpbenchmark/benchmarks/auctionmark/TestAuctionMarkWorker.java:63]: TL;DR: it made the error go away 😅

`this.workers` returns a `List<Worker<? extends BenchmarkModule>>` (i.e., a wildcard for any number of classes that extend `BenchmarkModule`).
Critically, it doesn't return `List<Workers<T extends BenchmarkModule>>`.
That means that `Worker w : this.workers` is technically an invalid cast.
But so is `Worker<AuctionMarkBenchmark>` even though that's what's actually in the list, despite what the type declaration says.
The problem is that the `this.workers` value actually gets populated from an overload of `BenchmarkModule` baseclass, `AuctionMarkBenchmark`, but that means that we'd wind up with a circular type definition (where the `AuctionMarkBenchmark extends BenchmarkModule` needs to return a `Worker<AuctionMarkBenchmark>`).
Anyways, that is quite complex so I wanted to push it off to another PR (if at all) and it turns out that by simply letting java infer the type via a `var` declaration, we avoid the issue (for now).
We're still stuck with the explicit cast after that though unfortunately.

File: src/main/java/com/oltpbenchmark/api/BenchmarkModule.java [hunk 4]
```diff
@@ -77,14 +77,15 @@ public abstract class BenchmarkModule {
     public BenchmarkModule(WorkloadConfiguration workConf) {
         this.workConf = workConf;
         this.dialects = new StatementDialects(workConf);
-        initClassLoader();
+        // setClassLoader();
+        this.classLoader = ClassLoader.getSystemClassLoader();
```
Comment (bpkroth) [inline src/main/java/com/oltpbenchmark/api/BenchmarkModule.java:81]: Adjusted to avoid a `this-escape` warning.

File: src/main/java/com/oltpbenchmark/benchmarks/auctionmark/AuctionMarkLoader.java [hunk 5]
```diff
@@ -46,7 +46,7 @@
  * @author pavlo
  * @author visawee
  */
-public class AuctionMarkLoader extends Loader<AuctionMarkBenchmark> {
+public final class AuctionMarkLoader extends Loader<AuctionMarkBenchmark> {
```
Comment (bpkroth) [inline src/main/java/com/oltpbenchmark/benchmarks/auctionmark/AuctionMarkLoader.java:49]: Many of these classes were converted to `final` to  avoid `this-escape` warnings in which non-final methods are used by the class to initialize the class.  In theory those could be overridden by a subclass to result in problematic initialization.

Since most of these classes are never expected to be subclassed, we simply avoid the error for now and mark them as `final`.

Comment (bpkroth) [reply src/main/java/com/oltpbenchmark/benchmarks/auctionmark/AuctionMarkLoader.java:49]: Added notes to this effect in the CONTRIBUTING.md file.
